### PR TITLE
interfaces: add an optional mount-host-font-cache plug attribute to the desktop interface

### PR DESCRIPTION
**Pull request by jhenstridge**
This PR adds a new optional plug attribute to the desktop interface that disables mounting of the system fontconfig cache.  If the attribute is not present, we treat it as `mount-host-font-cache: true` for backward compatibility.

The intent is that Snapcraft extensions that provide a private fontconfig cache could set this attribute to ensure that system fontconfig caches are not visible.

One thing I haven't included here but might make sense: should we mount an empty directory over `~/.fontconfig` and `~/.cache/fontconfig` as user mounts?  These could potentially cause compatibility problems, but won't usually be consulted by default because snaps run with a custom setting of $HOME.

### CODE REVIEW

**Comment by zyga** [comment]
Hey @jhenstridge 

About `~/.fontconfig` and `~/.cache/fontconfig`. While it is true HOME is altered and normally applications would not look there, IIRC destkop helpers at some point did explicitly copy the cache from those directories.

File: interfaces/builtin/desktop.go [hunk 1]
```diff
@@ -284,12 +287,34 @@ type desktopInterface struct {
 	commonInterface
 }
 
-func (iface *desktopInterface) fontconfigDirs() []string {
+func (iface *desktopInterface) mountFontCache(attribs interfaces.Attrer) (bool, error) {
```
Comment (zyga) [inline interfaces/builtin/desktop.go:290]: Nitpick, perhaps `shouldMountFontCache` so that the name doesn't imply actually mounting anything from snapd.

**Review by zyga — COMMENTED** [review]
Small naming nitpick. The code looks fine.

**Comment by jhenstridge** [comment]
As discussed on IRC, this PR will not affect any existing snaps since `mount-font-cache: true` is the default.  For the font caches in the home directory to be a problem, a snap would need to (a) add the upgrade their plug definitions, and (b) continue to use a version of the desktop-helpers that adds the home dir font caches to fontconfig's search path.

File: interfaces/builtin/desktop_test.go [hunk 2]
```diff
@@ -219,3 +219,42 @@ func (s *DesktopInterfaceSuite) TestStaticInfo(c *C) {
...
+
+func (s *DesktopInterfaceSuite) TestDisableMountFontCache(c *C) {
+	const mockSnapYaml = `name: desktop-snap
+version: 1.0
+plugs:
+  desktop:
+    mount-font-cache: false
+`
+	plug, plugInfo := MockConnectedPlug(c, mockSnapYaml, nil, "desktop")
+	c.Check(interfaces.BeforePreparePlug(s.iface, plugInfo), IsNil)
```
Comment (bboozzoo) [inline interfaces/builtin/desktop_test.go:231]: s/Check/Assert/ probably

File: interfaces/builtin/desktop.go [hunk 3]
```diff
@@ -284,12 +287,34 @@ type desktopInterface struct {
 	commonInterface
 }
 
-func (iface *desktopInterface) fontconfigDirs() []string {
+func (iface *desktopInterface) shouldMountFontCache(attribs interfaces.Attrer) (bool, error) {
+	value, ok := attribs.Lookup("mount-font-cache")
```
Comment (bboozzoo) [inline interfaces/builtin/desktop.go:291]: I'll be the first to start 2020 name bike shedding contest. Wondering about the name a bit. Maybe it should be called `private-font-cache` so that it's more about the actual intent of having a private cache rather than the actual action of mounting it which is an implementation detail after all.

Comment (jhenstridge) [reply interfaces/builtin/desktop.go:291]: My main concerns with the `private-font-cache` name is that (a) you could achieve a private cache without this setting with an appropriate fontconfig configuration, and (b) people might think it implies snapd will manage that cache some how (it's not going to do anything with the extra fc-cache binaries shiped with snapd, for instance).

I picked this name because it more closely reflects what it actually does.  With that said, I'm open to other suggestions though.

File: interfaces/builtin/desktop_test.go [hunk 4]
```diff
@@ -219,3 +219,42 @@ func (s *DesktopInterfaceSuite) TestStaticInfo(c *C) {
 func (s *DesktopInterfaceSuite) TestInterfaces(c *C) {
 	c.Check(builtin.Interfaces(), testutil.DeepContains, s.iface)
 }
+
+func (s *DesktopInterfaceSuite) TestDisableMountFontCache(c *C) {
+	const mockSnapYaml = `name: desktop-snap
+version: 1.0
+plugs:
+  desktop:
+    mount-font-cache: false
```
Comment (anonymouse64) [inline interfaces/builtin/desktop_test.go:228]: we also need test cases for this being true, and this being the empty string (though presumably there are pre-existing tests for the empty case so you don't need to add that if those already exist, I didn't check yet)

**Review by anonymouse64 — APPROVED** [review]
thanks for this, I'm divided on the name, so will let others bikeshed about that, but we need additional unit tests I think and the error message needs adjustment

**Review by anonymouse64 — CHANGES_REQUESTED** [review]
(meant to request changes until the additional unit tests are added)

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for this (but note probably needs @pedronis for the name of the attribute)

File: interfaces/builtin/desktop.go (hunk 3 above)
Comment (pedronis) [reply interfaces/builtin/desktop.go:291]: I think for claritiy this should be either be called:

* host-font-cache
* or more verbose  mount-host-font-cache

File: interfaces/builtin/desktop.go [hunk 5]
```diff
@@ -363,6 +396,11 @@ func (iface *desktopInterface) MountConnectedPlug(spec *mount.Specification, plu
 	return nil
 }
 
+func (iface *desktopInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	_, err := iface.shouldMountFontCache(plug)
```
Comment (pedronis) [inline interfaces/builtin/desktop.go:400]: thanks for this

**Review by pedronis — COMMENTED** [review]
thanks, a comment about the attribute name

File: interfaces/builtin/desktop.go (hunk 3 above)
Comment (jhenstridge) [reply interfaces/builtin/desktop.go:291]: I've renamed the attribute to `mount-host-font-cache`.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: interfaces/builtin/desktop_test.go (hunk 2 above)
Comment (jhenstridge) [reply interfaces/builtin/desktop_test.go:232]: I think I used Check here because the tests below it are not going to panic if this one fails.  It isn't making any modifications to the plugInfo that later parts of the test depend on.

**Review by anonymouse64 — APPROVED** [review]
still lgtm

**Comment by the-solipsist** [comment]
Hello folks. I'm just following up on https://forum.snapcraft.io/t/default-chromium-fonts-are-missing-after-snapd-has-been-updated/21651 

This commit seems to have been an important step towards getting the Chromium snap fixed.  Is there a timeline as to when the Chromium snap will be fixed and pushed to the snap store?

Currently, I need to run `sudo snap run --shell chromium -c 'FONTCONFIG_FILE="${SNAP_COMMON}/fontconfig/fonts.conf" "${SNAP_DESKTOP_RUNTIME}/usr/bin/fc-cache" --force --system-only --verbose'` each time the Chromium snap updates, and then I need to close all running programs, log out and log back in for the fonts to go back to normal.  As you can imagine, this is quite tedious.
